Create a Razor language server cohosted in Roslyn

davidwengier opened the pull request:
Part of https://github.com/dotnet/razor/issues/9519

This creates a basic LSP client/server for the "Razor" content type, inheriting from the existing Roslyn language server. Tried to keep the code as minimal as possible, as all of the endpoints will be in the Razor repo, but we need document sync in this repo, in order to access internals on RequestContext, and project context in this repo to match the data the RequestContext expects to receive. Some minor complexity in order to avoid exposing VS Protocol types in the API, to avoid dependency annoyances.

This also includes a very small portion of https://github.com/dotnet/roslyn/pull/70170, which is support for LSP requests against `TextDocument`s. Thank you @mgoertz-msft!

I've tested locally with local build Roslyn and Razor, but not used it in anger to try to implement a "real" endpoint in the Razor repo yet. Wanted to put this up early-ish in case there is a lot of feedback.

davidwengier commented on src/Features/LanguageServer/Protocol/Handler/ProjectContext/GetTextDocumentWithContextHandler.cs:
```diff
@@ -35,19 +35,20 @@ public GetTextDocumentWithContextHandler()
             Contract.ThrowIfNull(context.Solution);
 
-            // We specifically don't use context.Document here because we want multiple
-            var documents = context.Solution.GetDocuments(request.TextDocument.Uri);
```
Changing this to call `GetTextDocuments` seemed overly wasteful given only project info is needed, so I changed this handler a bit. Mainly wanted to avoid calling `GetDocument` and `GetAdditionalDocument` on every Id (and this endpoint is called a fair bit)

davidwengier commented on src/Tools/ExternalAccess/Razor/Microsoft.CodeAnalysis.ExternalAccess.Razor.csproj:
```diff
@@ -31,7 +31,6 @@
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.Razor" Key="$(RazorKey)" />
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServerClient.Razor" Key="$(RazorKey)" />
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServerClient.Razor.Test" Key="$(RazorKey)" />
-    <InternalsVisibleTo Include="Microsoft.VisualStudio.Mac.LanguageServices.Razor" Key="$(RazorKey)" />
```
Snuck this change in here, and the above repo uri change :)

davidwengier commented on src/Tools/ExternalAccess/Razor/Cohost/AbstractRazorRequestHandler.cs:
```diff
@@ -0,0 +1,78 @@
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost;
+
+internal abstract class AbstractRazorCohostRequestHandler<TRequestType, TResponseType> : ILspServiceRequestHandler<TRequestType, TResponseType>
```
Endpoints on the Razor side will inherit from one of these two abstract types.

davidwengier commented on src/Tools/ExternalAccess/RazorTest/Cohost/RazorCohostTests.cs:
```diff
@@ -0,0 +1,233 @@
+
+        protected override RazorTextDocumentIdentifier? GetRazorTextDocumentIdentifier(TextDocumentPositionParams request)
+        {
+            return new RazorTextDocumentIdentifier(request.TextDocument.Uri, (request.TextDocument as VSTextDocumentIdentifier)?.ProjectContext?.Id);
```
All of the endpoints in the Razor repo will have to do this, in particular the second parameter, in order to ensure project context flows correctly. Not too bad, and I couldn't see a way around it.

dibarbet reviewed (approved):
Generally fine with this

davidwengier commented on src/Features/LanguageServer/Protocol/WellKnownLspServerKinds.cs:
```diff
@@ -8,6 +8,11 @@ namespace Microsoft.CodeAnalysis.LanguageServer;
+    /// <summary>
+    /// Razor LSP server for Razor document requests (.razor and .cshtml files)
+    /// </summary>
+    RazorCohostServer,
```
Item 2 hadn't really occurred to me, but I guess we could do that. Without dynamic registration though, we probably can't do that behind a feature flag though, as it would just be controlled by the content type attribute on the server. Interesting idea though.

The rest is correct though. Essentially this lets us work on migration, find bugs, maybe get some benefit, but without being blocked on dynamic registration.

davidwengier commented on src/Tools/ExternalAccess/Razor/Cohost/RazorCohostLanguageClient.cs:
```diff
@@ -0,0 +1,88 @@
+    [Import(AllowDefault = true)] IRazorCohostLanguageClientActivationService? razorCohostLanguageClientActivationService = null)
+    : AbstractInProcLanguageClient(lspServiceProvider, globalOptions, lspLoggerFactory, threadingContext, exportProvider, middleLayer: null)
+{
+    private readonly IRazorCohostLanguageClientActivationService? _razorCohostLanguageClientActivationService = razorCohostLanguageClientActivationService;
```
No, this is just personal preference. If the Roslyn standard is to not use fields, I can remove.

dibarbet replied:
I think generally we remove them

davidwengier commented on src/Tools/ExternalAccess/Razor/Cohost/RazorLspServiceProvider.cs:
```diff
@@ -0,0 +1,21 @@
+[Export(typeof(RazorLspServiceProvider)), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class RazorLspServiceProvider(
```
Yep! Thats why I tried to make them do as little as possible

davidwengier commented:
@CyrusNajmabadi I heard a rumour you wanted to review this too?

I imagine there will be follow ups regardless, as I start implementing the Razor side

davidwengier commented on src/Tools/ExternalAccess/Razor/Cohost/RazorCohostLanguageClient.cs:
```diff
@@ -0,0 +1,88 @@
+            return base.ActivateAsync(cancellationToken);
+        }
+
+        return Task.FromResult<Connection?>(null);
```
Sadly, this doesn't work. The platform helpfully checks this for null and just immediately throws an exception 🤦‍♂️

https://dev.azure.com/devdiv/DevDiv/_git/VSLanguageServerClient?path=/src/product/RemoteLanguage/Impl/Infrastructure/MefClientProvider/RemoteLanguageClientInstance.cs&version=GBdevelop&line=528&lineEnd=529&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

So options are ClientName, or just return no capabilities. Thoughts @dibarbet ?

I'm guessing ClientName, since activating a server to nothing isn't great, but there is a slight downside with that, that we'd be unable to do progressive moves to this server, it would be all or nothing. I don't know how important that will be though.

dibarbet replied:
So I'm not super opposed to having a server with no capabilities.  We'll just have to make sure that the server isn't trying to do anything (e.g. sending refresh notifications).

As you mention it would allow you to migrate features one at a time which I think is fairly valuable.

davidwengier commented on src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractLanguageServer.cs:
```diff
@@ -353,6 +353,11 @@ internal TestAccessor(AbstractLanguageServer<TRequestContext> server)
             return null;
         }
 
+        internal Task<TResponse> ExecuteRequestAsync<TRequest, TResponse>(string methodName, TRequest request)
```
Oops, thanks, will move the None to the caller

davidwengier commented:
I've addressed all of the feedback, and am using this fine from the Razor repo. Just to confirm @dibarbet is this good to merge? Or do you want to hold off for a specific version etc.?

davidwengier commented:
Made a few more updates to this for issues found when implementing the first endpoint from Razor (in https://github.com/dotnet/razor/pull/9674). Would be good if you can take a look and re-approve @dibarbet, then I'll merge so I can get some packages going.

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/RequestContext.cs:
```diff
@@ -116,6 +116,43 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler;
+            }
+
+            // Explicitly throw for attempts to get a Document when only a TextDocument is available.
+            throw new InvalidOperationException();
```
We should put a message here so its distinguishable from the other invalid operaiton when the workspace is null

dibarbet commented on src/Features/LanguageServer/Protocol/WellKnownLspServerKinds.cs:
```diff
@@ -96,6 +104,7 @@ public static string GetContractName(this WellKnownLspServerKinds server)
     {
         return server switch
         {
+            WellKnownLspServerKinds.RazorCohostServer => "RazorLanguageServer",
```
Can we define the contract in the same protocol constants type, and also reference the protocol constants in the razor external access`Constants` class?

dibarbet commented on src/Tools/ExternalAccess/Razor/Cohost/RazorCohostDidChangeEndpoint.cs:
```diff
@@ -0,0 +1,56 @@
+        context.UpdateTrackedDocument(request.TextDocument.Uri, text);
+
+        // Razor can't handle this request because they don't have access to the RequestContext, but they might want to do something with it
+        if (didChangeHandler is not null)
```
Wondering here if we want a different mechanism to expose changes to Razor, if Razor needs to do special handling on didChange.  For example similar to the IOnInitialized services that get called by the base InitializedHandler.

Especially when we unify with the Roslyn server- I'm not sure we want to allow separate didChange handlers per language. 

Does the work Razor needs to do here require the queue to be blocked while its happening?

Same for open/close

davidwengier replied:
All good thoughts, and I see this as definitely a point-in-time thing.

1. When we have dynamic registration, yes I would expect the "normal" text document sync handlers work for all languages and we'd remove these.
2. When Razor can pull on the source generator for our .g.cs files then we won't need to be notified about changes or opens.

For now though, we need something here, as the Razor side that responds to this is effectively unchanged from our existing code, we kick off a compile of the razor file as it changes.

davidwengier replied:
Actually, even when we can use the source generator and even use the Roslyn API rather than LSP to ask questions about it etc, we'd still need some way to get notified of open and change of Razor files, and run code while the queue is blocked, I order to make sure the html virtual buffers are up to date. Maybe some way to hook into the LspWorkspaceManager? Haven't really thought about the specifics of that far ahead yet.

dibarbet replied:
Hmm.  Probably depends on what needs to happen.  I wouldn't be opposed to something like IOnInitialized where we expose a hook inside the handler.  I think putting it there makes it more clear that its a blocking hook (than putting in LspWorkspaceManager0

davidwengier replied:
I'm missing something. Isn't that exactly what this is currently doing?

dibarbet replied:
What I meant was putting code inside the main DidChangeHandler that allows for arbitrary additional callers, instead of duplicating teh handler for Razor and having a special Razor hook.
e.g. https://github.com/dotnet/roslyn/blob/main/src/Features/LanguageServer/Protocol/Handler/ServerLifetime/InitializedHandler.cs#L27

We'd have a generic hook that razor could implement.  However I think you need to export a separate Razor didChange handler already - so I'm fine if we do that later when we merge the servers.

davidwengier replied:
Oh yep, can definitely follow up with that once we merge. The existing handlers won't activate in this server though, due to the export attribute being used.

davidwengier commented on src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractLanguageServer.cs:
```diff
@@ -353,6 +353,11 @@ internal TestAccessor(AbstractLanguageServer<TRequestContext> server)
             return null;
         }
 
+        internal Task<TResponse> ExecuteRequestAsync<TRequest, TResponse>(string methodName, TRequest request, CancellationToken cancellationToken)
```
The request here is still part of the clients request ordering, this just effectively adds something to the queue. The existing tests don't need to do this, because they create a `TestLspServer` that exposes this, but we need the RazorCohost server. I could change to call Invoke on the JsonRpc if you prefer that? But otherwise I'm not entirely sure what the alternative is

dibarbet replied:
Oh my bad, I didn't see it was in the test accessor!

davidwengier commented:
@dibarbet any reason not to merge this?
